lib/remove.js

```diff
@@ -16,6 +16,6 @@ module.exports = remove
  */
 function remove (state, prefix, objectsOrIds, change) {
   return Array.isArray(objectsOrIds)
-    ? updateMany(state, objectsOrIds.map(markAsDeleted.bind(null, change)), null, prefix)
-    : updateOne(state, markAsDeleted(change, objectsOrIds), null, prefix)
+    ? updateMany(state, objectsOrIds.map(markAsDeleted.bind(null, state, change)), null, prefix)
+    : updateOne(state, markAsDeleted(state, change, objectsOrIds), null, prefix)
```

Comment:
`markAsDeleted` runs `changeObject` which is now asynchronous, so I think that won’t work? We have to make `markAsDeleted` async, too, right?

Reply:
`changeObject` is asynchronous only when we pass it's `validateChange` argument not as `false`
`markAsDeleted` is calling out `changeObject` with `validateChange` set to `false` , therefore it will be synchronous

Reply:
got it. But this is really confusing, we gotta make this simpler somehow. I’m still catching up with emails & GitHub notifications from last two weeks, I hope to be able to help out soon

Reply:
One thing which we can do to simplify this would be to make `markAsDeleted` an asynchronous utility. The only overhead would be that the deletion flow will undergo two validations. One after marking for deletion and another before updating the database.

Reply:
> make markAsDeleted an asynchronous utility

+1 on that

> One after marking for deletion and another before updating the database

Can you point me to the two places right now using an example? It’ll help me a lot, I’m all over different projects right now <3

Reply:
`lib/util/change-object.js` was modified to validate the "changes" to a `doc` before taking any action. So, basically whenever you "change" an object (in reference to the current discussion, either during update or when you're marking for deletion), we will validate the "changes" and hence it happens twice for deletion flow in the following order :
* mutating `doc` to  add key for "marking it for deletion" (through `changeObject`)
* when we call the update method, we'll execute `changeObject` downstream and hence, trigger the validation again.

Reply:
thanks! I’m still catching up from past two weeks of travel, I hope to find time to look into this proper. 

From what you wrote, it sounds to me that we just shouldn’t use the `changeObject` method to mark a document as deleted, so it only gets run twice?

Reply:
Sorry for the delay in responding to your comment @gr2m . I was out sick.

> From what you wrote, it sounds to me that we just shouldn’t use the changeObject method to mark a document as deleted, so it only gets run twice?

Yeah, if we want to avoid validating twice for the deletion flow.

Another existing problem which comes to mind is when update fails for deletion requests, why don't we delete the key we added to the local document to mark it "to be deleted" ?

Reply:
> I was out sick.

Glad you are better, good to have you back :)

When the update fails the validation, then the document should not be changed at all, including the `_deleted: true` property, no?

Reply:
Why so though? Won't that make the document inconsistent? Considering, the field marks that it has been deleted but actually due to failing the validation/update, the doc is still present in the DB.

Reply:
I mean setting the `_deleted` property in `store.remove` is part of the update. If I do `store.remove('id123', {foo: 'bar'})` and the validation fails because the `foo` property is not allowed, then the whole remove fails. I think we have a misunderstanding somehow :)

Reply:
Shouldn't removing the `_deleted` key be a part of rollback if remove fails is what I'm asking? That isn't being done at the moment.

Reply:
I'm not sure what  you mean with "remove fails". How can a remove fail? If it fails due to the validation, nothing is written to the store, it fails before anything is written

Reply:
Nothing is written to the store, but I guess the local document has been mutated.

Reply:
A got it. Yeah that should not happen. Let's add a test for `store.update` where validation fails and check that the object passed to store.update didn't get mutated?

Reply:
I'll put up some changes and we'll take it from there!

Reply:
awesome thanks Mayank! Sorry for not being more responsive, I'll be around more once Offline Camp is done next weekend 🙏

Reply:
It's alright! I've been a bit busy myself 🙂